README.md
# docker-junit-rule

A junit rule to run docker containers

[![Build Status](https://travis-ci.org/geowarin/docker-junit-rule.svg)](https://travis-ci.org/geowarin/docker-junit-rule)

## Usage

Example for rabbitMQ:

```java
import com.github.geowarin.junit.DockerRule;
import com.rabbitmq.client.ConnectionFactory;
import org.junit.ClassRule;
import org.junit.Test;

public class RabbitIntegrationTest {

  @ClassRule
  public static DockerRule rabbitRule =
    DockerRule.builder()
      .image("rabbitmq:management")
      .ports("5672")
//      .waitForPort("5672/tcp")
      .waitForLog("Server startup complete")
      .build();

  @Test
  public void testConnectsToDocker() throws Exception {
    ConnectionFactory factory = new ConnectionFactory();
    factory.setHost(rabbitRule.getDockerHost());
    factory.setPort(rabbitRule.getHostPort("5672/tcp"));
    factory.newConnection();
  }
}
```

## Installation

The library is available on jcenter

### Maven

Add the following to your `pom.xml`:

```xml
<repositories>
    <repository>
        <snapshots>
            <enabled>false</enabled>
        </snapshots>
        <id>central</id>
        <name>bintray</name>
        <url>http://jcenter.bintray.com</url>
    </repository>
</repositories>

...

<dependency>
    <groupId>com.github.geowarin</groupId>
    <artifactId>docker-junit-rule</artifactId>
    <version>1.1.0</version>
    <scope>test</scope>
</dependency>
```

### Gradle

Add the following to your `build.gradle`:

```groovy
repositories {
  jcenter()
}

dependencies {
  testCompile 'com.github.geowarin:docker-junit-rule:1.1.0'
}
```

## Principle

Uses https://github.com/spotify/docker-client to connect to the docker daemon API.

Tested with docker-for-mac and travis runs it on linux.
If it does not work with docker-for-windows, please open a PR ;)

## Licence

MIT


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 1.2.0

- Add `isLocalImage` to the builder to specify image which should not be pulled from the registry (#4 by @fagossa)
- Upgrade to com.spotify:docker-client:8.9.1 
⚠️ We no longer use the shaded version because of https://github.com/spotify/docker-client/issues/900
- Fix default environment (it now works with docker for mac)

## 1.1.0

- Add a builder to remove the need for inheritance
- Add a waitForLog directive

## 1.0.2
> 2016-01-06

- Include guava in the dependencies

## 1.0.1
> 2016-01-06

- Use the shaded jar of the docker-client

## 1.0.0
> 2016-01-06

- Initial release


src/test/java/rules/SocketUtil.java
package rules;

import java.io.IOException;
import java.net.ServerSocket;

/**
 * Created by geowarin on 06/01/16.
 */
public class SocketUtil {
    static int findFreePort() {
        ServerSocket socket = null;
        try {
            socket = new ServerSocket(0);
            return socket.getLocalPort();
        } catch (IOException e) {
        } finally {
            if (socket != null) {
                try {
                    socket.close();
                } catch (IOException e) {
                }
            }
        }
        return -1;
    }
}


src/test/java/rules/DockerContainerRule.java
package rules;

import com.spotify.docker.client.*;
import com.spotify.docker.client.messages.*;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.junit.rules.ExternalResource;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.net.URI;
import java.nio.channels.SocketChannel;
import java.nio.file.Paths;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * JUnit rule starting a docker container before the test and killing it
 * afterwards.
 *
 * Uses spotify/docker-client.
 *
 * Adapted from https://gist.github.com/mosheeshel/c427b43c36b256731a0b
 *
 * author: Geoffroy Warin (geowarin.github.io)
 */
public class DockerContainerRule extends ExternalResource {
    protected final Log logger = LogFactory.getLog(getClass());
    public static final String DOCKER_MACHINE_SERVICE_URL = "https://192.168.99.100:2376";

    private final DockerClient dockerClient;
    private ContainerCreation container;
    private Map<String, List<PortBinding>> ports;

    public DockerContainerRule(String imageName) {
        this(imageName, new String[0], null);
    }

    public DockerContainerRule(String imageName, String[] ports) {
        this(imageName, ports, null);
    }

    /**
     * @param imageName The name of the docker image to use
     * @param ports     The ports that will be open on the container. Will automatically assign random ports on the host
     * @param cmd       override the default container cmd
     */
    public DockerContainerRule(String imageName, String[] ports, String cmd) {
        dockerClient = createDockerClient();
        ContainerConfig containerConfig = createContainerConfig(imageName, ports, cmd);

        try {
            dockerClient.pull(imageName);
            container = dockerClient.createContainer(containerConfig);
        } catch (DockerException | InterruptedException e) {
            throw new IllegalStateException(e);
        }
    }

    @Override
    protected void before() throws Throwable {
        super.before();
        dockerClient.startContainer(container.id());
        ContainerInfo info = dockerClient.inspectContainer(container.id());
        ports = info.networkSettings().ports();
    }

    @Override
    protected void after() {
        super.after();
        try {
            dockerClient.killContainer(container.id());
            dockerClient.removeContainer(container.id(), true);
            dockerClient.close();
        } catch (DockerException | InterruptedException e) {
            throw new RuntimeException("Unable to stop/remove docker container " + container.id(), e);
        }
    }

    /**
     * Utility method to get the docker host.
     * Can be different from localhost if using docker-machine
     */
    public String getDockerHost() {
        return dockerClient.getHost();
    }

    /**
     * Utility method to ensure a container is started
     */
    public void waitForPort(int port) {
        waitForPort(port, 10000);
    }

    /**
     * Utility method to ensure a container is started
     */
    public void waitForPort(int port, long timeoutInMillis) {
        SocketAddress address = new InetSocketAddress(getDockerHost(), port);
        long totalWait = 0;
        while (true) {
            try {
                SocketChannel.open(address);
                return;
            } catch (IOException e) {
                try {
                    Thread.sleep(100);
                    totalWait += 100;
                    if (totalWait > timeoutInMillis) {
                        throw new IllegalStateException("Timeout while waiting for port " + port);
                    }
                } catch (InterruptedException ie) {
                    throw new IllegalStateException(ie);
                }
            }
        }
    }

    private DockerClient createDockerClient() {
        if (isUnix() || System.getenv("DOCKER_HOST") != null) {
            try {
                return DefaultDockerClient.fromEnv().build();
            } catch (DockerCertificateException e) {
                System.err.println(e.getMessage());
            }
        }

        logger.info("Could not create docker client from the environment. Assuming docker-machine environment with url " + DOCKER_MACHINE_SERVICE_URL);
        DockerCertificates dockerCertificates = null;
        try {
            String userHome = System.getProperty("user.home");
            dockerCertificates = new DockerCertificates(Paths.get(userHome, ".docker/machine/certs"));
        } catch (DockerCertificateException e) {
            System.err.println(e.getMessage());
        }
        return DefaultDockerClient.builder()
                .uri(URI.create(DOCKER_MACHINE_SERVICE_URL))
                .dockerCertificates(dockerCertificates)
                .build();
    }

    private ContainerConfig createContainerConfig(String imageName, String[] ports, String cmd) {
        Map<String, List<PortBinding>> portBindings = new HashMap<>();
        for (String port : ports) {
            List<PortBinding> hostPorts = Collections.singletonList(PortBinding.randomPort("0.0.0.0"));
            portBindings.put(port, hostPorts);
        }

        HostConfig hostConfig = HostConfig.builder()
                .portBindings(portBindings)
                .build();

        ContainerConfig.Builder configBuilder = ContainerConfig.builder()
                .hostConfig(hostConfig)
                .image(imageName)
                .networkDisabled(false)
                .exposedPorts(ports);

        if (cmd != null) {
            configBuilder = configBuilder.cmd(cmd);
        }
        return configBuilder.build();
    }

    public int getHostPort(String containerPort) {
        List<PortBinding> portBindings = ports.get(containerPort);
        if (portBindings.isEmpty()) {
            return -1;
        }
        return Integer.parseInt(portBindings.get(0).hostPort());
    }

    private static boolean isUnix() {
        String os = System.getProperty("os.name").toLowerCase();
        return os.contains("nix") || os.contains("nux") || os.contains("aix");
    }
}

src/test/java/rules/RabbitContainerRule.java
package rules;

import com.github.geowarin.junit.DockerContainerRule;

public class RabbitContainerRule extends DockerContainerRule {
  public static final String RABBIT_CONTAINER_IMAGE_NAME = "rabbitmq:management";

  public RabbitContainerRule() {
    super(RABBIT_CONTAINER_IMAGE_NAME, new String[]{"5672", "61613", "15672"});
  }

  @Override
  protected void before() throws Throwable {
    super.before();
    // wait for container to boot
    waitForPort(getRabbitServicePort());
  }

  public int getRabbitServicePort() {
    return getHostPort("5672/tcp");
  }

  public int getRabbitManagementPort() {
    return getHostPort("15672/tcp");
  }
}


src/main/java/producer/RabbitProducer.java
package producer;

import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import com.rabbitmq.client.MessageProperties;

import java.io.IOException;

public class RabbitProducer {
    private Channel channel;

    public RabbitProducer(String host, int port) throws IOException {
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost(host);
        factory.setPort(port);
        Connection connection = factory.newConnection();
        channel = connection.createChannel();
        this.channel.queueDeclare("greetings", false, false, true, null);
    }

    public void produce() throws IOException {
        channel.basicPublish("greetings", "", MessageProperties.PERSISTENT_TEXT_PLAIN, "hello".getBytes());
    }
}


src/main/java/com/github/geowarin/junit/DockerRule.java
package com.github.geowarin.junit;

import com.spotify.docker.client.DefaultDockerClient;
import com.spotify.docker.client.DockerClient;
import com.spotify.docker.client.LogMessage;
import com.spotify.docker.client.LogStream;
import com.spotify.docker.client.exceptions.DockerCertificateException;
import com.spotify.docker.client.exceptions.DockerException;
import com.spotify.docker.client.messages.*;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.junit.rules.ExternalResource;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.SocketChannel;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static com.spotify.docker.client.DockerClient.LogsParam.follow;
import static com.spotify.docker.client.DockerClient.LogsParam.stdout;
import static com.spotify.docker.client.DockerClient.RemoveContainerParam.removeVolumes;

/**
 * <p>
 * JUnit rule starting a docker container before the test and killing it
 * afterwards.
 * </p>
 * <p>
 * Uses spotify/docker-client.
 * Adapted from https://gist.github.com/mosheeshel/c427b43c36b256731a0b
 * </p>
 * author: Geoffroy Warin (geowarin.github.io)
 */
public class DockerRule extends ExternalResource {
  protected final Log logger = LogFactory.getLog(getClass());
  public static final String DOCKER_MACHINE_SERVICE_URL = "https://192.168.99.100:2376";

  private final DockerClient dockerClient;
  private ContainerCreation container;
  private Map<String, List<PortBinding>> ports;
  private DockerRuleParams params;

  public static DockerRuleBuilder builder() {
    return new DockerRuleBuilder();
  }

  DockerRule(DockerRuleParams params) {
    this.params = params;
    dockerClient = createDockerClient();
    ContainerConfig containerConfig = createContainerConfig(params.imageName, params.ports, params.cmd);

    try {
      if (!params.isLocalImage) {
        dockerClient.pull(params.imageName);
      }
      container = dockerClient.createContainer(containerConfig);
    } catch (DockerException | InterruptedException e) {
      throw new IllegalStateException(e);
    }
  }

  @Override
  protected void before() throws Throwable {
    super.before();
    dockerClient.startContainer(container.id());
    ContainerInfo info = dockerClient.inspectContainer(container.id());
    ports = info.networkSettings().ports();

    if (params.portToWaitOn != null) {
      waitForPort(getHostPort(params.portToWaitOn), params.waitTimeout);
    }

    if (params.logToWait != null) {
      waitForLog(params.logToWait);
    }
  }

  @Override
  protected void after() {
    super.after();
    try {
      dockerClient.killContainer(container.id());
      dockerClient.removeContainer(container.id(), removeVolumes());
      dockerClient.close();
    } catch (DockerException | InterruptedException e) {
      throw new RuntimeException("Unable to stop/remove docker container " + container.id(), e);
    }
  }

  /**
   * Utility method to get the docker host.
   * Can be different from localhost if using docker-machine
   *
   * @return The current docker host
   */
  public String getDockerHost() {
    return dockerClient.getHost();
  }

  protected void waitForPort(int port, long timeoutInMillis) {
    SocketAddress address = new InetSocketAddress(getDockerHost(), port);
    long totalWait = 0;
    while (true) {
      try {
        SocketChannel.open(address);
        return;
      } catch (IOException e) {
        try {
          Thread.sleep(100);
          totalWait += 100;
          if (totalWait > timeoutInMillis) {
            throw new IllegalStateException("Timeout while waiting for port " + port);
          }
        } catch (InterruptedException ie) {
          throw new IllegalStateException(ie);
        }
      }
    }
  }

  protected DockerClient createDockerClient() {
    try {
      return DefaultDockerClient.fromEnv().build();
    } catch (DockerCertificateException e) {
      throw new IllegalStateException("Could not create docker client from environement", e);
    }
  }

  protected ContainerConfig createContainerConfig(String imageName, String[] ports, String cmd) {
    Map<String, List<PortBinding>> portBindings = new HashMap<>();
    for (String port : ports) {
      List<PortBinding> hostPorts = Collections.singletonList(PortBinding.randomPort("0.0.0.0"));
      portBindings.put(port, hostPorts);
    }

    HostConfig hostConfig = HostConfig.builder()
      .portBindings(portBindings)
      .build();

    ContainerConfig.Builder configBuilder = ContainerConfig.builder()
      .hostConfig(hostConfig)
      .image(imageName)
      .networkDisabled(false)
      .exposedPorts(ports);

    if (cmd != null) {
      configBuilder = configBuilder.cmd(cmd);
    }
    return configBuilder.build();
  }

  public int getHostPort(String containerPort) {
    List<PortBinding> portBindings = ports.get(containerPort);
    if (portBindings.isEmpty()) {
      return -1;
    }
    return Integer.parseInt(portBindings.get(0).hostPort());
  }

  protected void waitForLog(String messageToMatch) throws DockerException, InterruptedException, UnsupportedEncodingException {
    LogStream logs = dockerClient.logs(container.id(), follow(), stdout());
    String log;
    do {
      LogMessage logMessage = logs.next();
      ByteBuffer buffer = logMessage.content();
      byte[] bytes = new byte[buffer.remaining()];
      buffer.get(bytes);
      log = new String(bytes);
    } while (!log.contains(messageToMatch));
  }
}


src/main/java/com/github/geowarin/junit/DockerRuleBuilder.java
package com.github.geowarin.junit;

import com.spotify.docker.client.messages.ContainerConfig;

public class DockerRuleBuilder {
  private final DockerRuleParams params = new DockerRuleParams();

  /**
   * @param imageName The name of the docker image to download
   * @return The builder
   */
  public DockerRuleBuilder image(String imageName) {
    params.imageName = imageName;
    return this;
  }

  /**
   * <p>
   * List the container ports that you would like to open in the container.
   * They will be bound on your host on random ports.
   * </p>
   * <p>
   * To know which ports are used on your host use DockerRule#getHostPort(String).
   * Example:
   * </p>
   * <pre>
   * {@code
   * myRule.getHostPort("80/tcp")
   * }
   * </pre>
   *
   * @param ports The ports to open on the container
   * @return The builder
   * @see DockerRule#getHostPort(String)
   */
  public DockerRuleBuilder ports(String... ports) {
    params.ports = ports;
    return this;
  }

  public DockerRuleBuilder cmd(String cmd) {
    params.cmd = cmd;
    return this;
  }

  /**
   * Utility method to ensure a container is started
   *
   * @param portToWaitOn The port to wait on
   * @return The builder
   */
  public DockerRuleBuilder waitForPort(String portToWaitOn) {
    return waitForPort(portToWaitOn, 10000);
  }

  /**
   * Utility method to ensure a container is started
   *
   * @param portToWaitOn    The port to wait on
   * @param timeoutInMillis Maximum waiting time in milliseconds
   * @return The builder
   */
  public DockerRuleBuilder waitForPort(String portToWaitOn, int timeoutInMillis) {
    params.portToWaitOn = portToWaitOn;
    params.waitTimeout = timeoutInMillis;
    return this;
  }

  /**
   * @param logToWait The log message to wait.
   *                  This will stop blocking as soon as the docker logs contains that string
   * @return The builder
   */
  public DockerRuleBuilder waitForLog(String logToWait) {
    params.logToWait = logToWait;
    return this;
  }

  public DockerRuleBuilder config(ContainerConfig config) {
    params.config = config;
    return this;
  }

  public DockerRuleBuilder dockerHost(String dockerHost) {
    params.dockerHost = dockerHost;
    return this;
  }

  public DockerRuleBuilder dockerCertPath(String dockerCertPath) {
    params.dockerCertPath = dockerCertPath;
    return this;
  }

  public DockerRule build() {
    return new DockerRule(params);
  }
}


src/main/java/com/github/geowarin/junit/DockerContainerRule.java
package com.github.geowarin.junit;

import com.spotify.docker.client.*;
import com.spotify.docker.client.messages.*;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.junit.rules.ExternalResource;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.net.URI;
import java.nio.channels.SocketChannel;
import java.nio.file.Paths;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * JUnit rule starting a docker container before the test and killing it
 * afterwards.
 *
 * <p>
 * Uses spotify/docker-client.
 * Adapted from https://gist.github.com/mosheeshel/c427b43c36b256731a0b
 * </p>
 *
 * author: Geoffroy Warin (geowarin.github.io)
 */
public class DockerContainerRule extends ExternalResource {
  protected final Log logger = LogFactory.getLog(getClass());
  public static final String DOCKER_MACHINE_SERVICE_URL = "https://192.168.99.100:2376";

  private final DockerClient dockerClient;
  private ContainerCreation container;
  private Map<String, List<PortBinding>> ports;

  public DockerContainerRule(String imageName) {
    this(imageName, new String[0], null);
  }

  public DockerContainerRule(String imageName, String[] ports) {
    this(imageName, ports, null);
  }

  /**
   * @param imageName The name of the docker image to use
   * @param ports     The ports that will be open on the container. Will automatically assign random ports on the host
   * @param cmd       override the default container cmd
   */
  public DockerContainerRule(String imageName, String[] ports, String cmd) {
    dockerClient = createDockerClient();
    ContainerConfig containerConfig = createContainerConfig(imageName, ports, cmd);

    try {
      dockerClient.pull(imageName);
      container = dockerClient.createContainer(containerConfig);
    } catch (DockerException | InterruptedException e) {
      throw new IllegalStateException(e);
    }
  }

  @Override
  protected void before() throws Throwable {
    super.before();
    dockerClient.startContainer(container.id());
    ContainerInfo info = dockerClient.inspectContainer(container.id());
    ports = info.networkSettings().ports();
  }

  @Override
  protected void after() {
    super.after();
    try {
      dockerClient.killContainer(container.id());
      dockerClient.removeContainer(container.id(), true);
      dockerClient.close();
    } catch (DockerException | InterruptedException e) {
      throw new RuntimeException("Unable to stop/remove docker container " + container.id(), e);
    }
  }

  /**
   * Utility method to get the docker host.
   * Can be different from localhost if using docker-machine
   *
   * @return The current docker host
   */
  public String getDockerHost() {
    return dockerClient.getHost();
  }

  /**
   * Utility method to ensure a container is started
   *
   * @param port The port to wait on
   */
  public void waitForPort(int port) {
    waitForPort(port, 10000);
  }

  /**
   * Utility method to ensure a container is started
   * @param port The port to wait on
   * @param timeoutInMillis Maximum waiting time in milliseconds
   */
  public void waitForPort(int port, long timeoutInMillis) {
    SocketAddress address = new InetSocketAddress(getDockerHost(), port);
    long totalWait = 0;
    while (true) {
      try {
        SocketChannel.open(address);
        return;
      } catch (IOException e) {
        try {
          Thread.sleep(100);
          totalWait += 100;
          if (totalWait > timeoutInMillis) {
            throw new IllegalStateException("Timeout while waiting for port " + port);
          }
        } catch (InterruptedException ie) {
          throw new IllegalStateException(ie);
        }
      }
    }
  }

  private DockerClient createDockerClient() {
    if (isUnix() || System.getenv("DOCKER_HOST") != null) {
      try {
        return DefaultDockerClient.fromEnv().build();
      } catch (DockerCertificateException e) {
        System.err.println(e.getMessage());
      }
    }

    logger.info("Could not create docker client from the environment. Assuming docker-machine environment with url " + DOCKER_MACHINE_SERVICE_URL);
    DockerCertificates dockerCertificates = null;
    try {
      String userHome = System.getProperty("user.home");
      dockerCertificates = new DockerCertificates(Paths.get(userHome, ".docker/machine/certs"));
    } catch (DockerCertificateException e) {
      System.err.println(e.getMessage());
    }
    return DefaultDockerClient.builder()
      .uri(URI.create(DOCKER_MACHINE_SERVICE_URL))
      .dockerCertificates(dockerCertificates)
      .build();
  }

  private ContainerConfig createContainerConfig(String imageName, String[] ports, String cmd) {
    Map<String, List<PortBinding>> portBindings = new HashMap<>();
    for (String port : ports) {
      List<PortBinding> hostPorts = Collections.singletonList(PortBinding.randomPort("0.0.0.0"));
      portBindings.put(port, hostPorts);
    }

    HostConfig hostConfig = HostConfig.builder()
      .portBindings(portBindings)
      .build();

    ContainerConfig.Builder configBuilder = ContainerConfig.builder()
      .hostConfig(hostConfig)
      .image(imageName)
      .networkDisabled(false)
      .exposedPorts(ports);

    if (cmd != null) {
      configBuilder = configBuilder.cmd(cmd);
    }
    return configBuilder.build();
  }

  public int getHostPort(String containerPort) {
    List<PortBinding> portBindings = ports.get(containerPort);
    if (portBindings.isEmpty()) {
      return -1;
    }
    return Integer.parseInt(portBindings.get(0).hostPort());
  }

  private static boolean isUnix() {
    String os = System.getProperty("os.name").toLowerCase();
    return os.contains("nix") || os.contains("nux") || os.contains("aix");
  }
}
